### PR TITLE
ModuleFactory

### PR DESCRIPTION
**Pull request by Atticus1806**
As discussed today moved out of #8

### CODE REVIEW

File: i6_models/config.py [hunk 1]
```diff
@@ -17,6 +17,8 @@ class ExampleModule(Module):
 from __future__ import annotations
 from dataclasses import dataclass, fields
 import typeguard
+from torch import nn
+from typing import TypeVar, Type
```
Comment (SimBe195) [inline i6_models/config.py:21]:
```suggestion
from typing import Callable, Generic, TypeVar
```

File: i6_models/config.py [hunk 2]
```diff
@@ -51,3 +53,22 @@ def _validate_types(self) -> None:
...
     def __post_init__(self) -> None:
         self._validate_types()
+
+
+ConfigType = TypeVar("ConfigType", bound=ModelConfiguration)
+ModuleType = TypeVar("ModuleType", bound=nn.Module)
+
+
+@dataclass
+class SubassemblyFactory:
```
Comment (SimBe195) [inline i6_models/config.py:63]:
```suggestion
class SubassemblyFactory(Generic[ConfigType, ModuleType]):
```

File: i6_models/config.py [hunk 3]
```diff
@@ -51,3 +53,22 @@ def _validate_types(self) -> None:
...
+    """
+    Dataclass for a combination of a Subassembly/Part and the corresponding configuration.
+    Also provides a function to construct the corresponding object through this dataclass
+    """
+
+    module_class: Type[ModuleType]
```
Comment (SimBe195) [inline i6_models/config.py:69]:
```suggestion
    module_class: Callable[[ConfigType], ModuleType]
```

**Review by SimBe195 — CHANGES_REQUESTED** [review]

File: i6_models/config.py (hunk 3 above)
Comment (Atticus1806) [reply i6_models/config.py:69]: I am not sure about this. I thought `Type` was meant for class objects.

File: i6_models/config.py (hunk 2 above)
Comment (Atticus1806) [reply i6_models/config.py:63]: We can add this, but I have to admit even after reading up on it I dont fully understand the benefits over just having the parameters typed. Also is the order within the generic relevant? Then the types would need to be swapped right?

File: i6_models/config.py (hunk 3 above)
Comment (SimBe195) [reply i6_models/config.py:69]: Yes but each class is also a callable due to its constructor. This typehint ensures that the constructor takes a single parameter of type `ConfigType` so that calling `self.module_class(self.cfg)` is valid.

File: i6_models/config.py (hunk 2 above)
Comment (SimBe195) [reply i6_models/config.py:63]: If you want to use generic types inside a class you usually declare it as a `Generic` depending on these generic types. Having this "connects" the generic types used inside. For example `ModuleType` used in `module_class` is the same as the `ModuleType` returned by the `__call__` function. So for example if you set `module_class = ConformerEncoderV1` somewhere type checkers will know that the `__call__` method will return a `ConformerEncoderV1` instance.

> Also is the order within the generic relevant?

No, I don't believe the order is relevant.

Comment (michelwi) [reply i6_models/config.py:63]: After reading a bit myself, I think the advantage is that we could specify which sort of subassemblies we would like to allow. e.g.

```
# define a metaclass for Norms
@dataclass
class NormAssemblyConfiguration(ModelConfiguration):
  pass

class NormAssembly(torch.nn.Module):
  pass

# define some norms
@dataclass
class BatchNormAssemblyConfiguration(NormAssemblyConfiguration):
  pass

class BatchNormAssembly(NormAssembly):
  pass

@dataclass
class LayerNormAssemblyConfiguration(NormAssemblyConfiguration):
  pass

class LayerNormAssembly(NormAssembly):
  pass

# some non norm related Assembly
class UnrelatedAssemblyConfiguration(ModelConfiguration):
  pass

class UnrelatedAssembly(torch.nn.Module):
  pass

# define a custom AssemblyConfig, that should accept any norm, but only norms
@dataclass
class MyAssemblyConfig(ModelConfiguration):
   norm: SubassemblyFactory[NormAssemblyConfiguration, NormAssembly]

# test the new config class
c1 = MyAssemblyConfig(
    norm=SubassemblyFactory(
        module_class=BatchNormAssembly,
        cfg=BatchNormAssemblyConfiguration()
    )
)  # correct typing

c2 = MyAssemblyConfig(
    norm=SubassemblyFactory(
        module_class=UnrelatedAssembly,
        cfg=UnrelatedAssemblyConfiguration()
    )
)  # valid python, but should give typing error
```
see my test here
image


> Also is the order within the generic relevant?

you should match the order when typing your own configs. In the example above
```
norm: SubassemblyFactory[NormAssemblyConfiguration, NormAssembly]  # same order
norm: SubassemblyFactory[NormAssembly, NormAssemblyConfiguration]  # wrong order
```

> Then the types would need to be swapped right?

this is a matter of taste. I would agree, but there is no technical reason the order here should match the order of arguments in the SubassemblyFactory

Comment (michelwi) [reply i6_models/config.py:63]: > If you want to use generic types inside a class you usually declare it as a Generic depending on these generic types. Having this "connects" the generic types used inside. 

I thought the fact that `ModuleType` is a `TypeVar` already did this, but apparently not.
image
(Type variable "ModuleType" has no meaning in this context)

However, if you move away from dataclasses, then you do not need to make the class a Generic to use TypeVars to enforce consistency:
image
("LayerNormAssemblyConfiguration" is incompatible with "BatchNormAssemblyConfiguration")

Comment (SimBe195) [reply i6_models/config.py:63]: > However, if you move away from dataclasses, then you do not need to make the class a Generic to use TypeVars to enforce consistency

This example doesn't work however when you add `def __call__(self) -> ModuleType:` because the system doesn't know that the `ModuleType` from `__init__` is the same `ModuleType` as in `__call__` if you don't inherit from `Generic`.

Comment (michelwi) [reply i6_models/config.py:63]: > This example doesn't work however when  [...]

Indeed it does not. 😅 
Thanks for the clarification and counterexample.

**Review by SimBe195 — APPROVED** [review]

**Review by michelwi — APPROVED** [review]

File: i6_models/config.py [hunk 4]
```diff
@@ -51,3 +53,22 @@ def _validate_types(self) -> None:
...
+
+
+@dataclass
+class SubassemblyFactory(Generic[ConfigType, ModuleType]):
+    """
+    Dataclass for a combination of a Subassembly/Part and the corresponding configuration.
+    Also provides a function to construct the corresponding object through this dataclass
+    """
+
+    module_class: Callable[[ConfigType], ModuleType]
```
Comment (albertz) [inline i6_models/config.py:69]: I think this should be:
```suggestion
    module_class: Type[ModuleType]
```
You want to be able to have code like:
```python
assert isinstance(obj, opt.module_class)
```
So it would be good to have that guarantee that this `module_class` is really the class, and not any random callable.

File: i6_models/config.py [hunk 5]
```diff
@@ -51,3 +53,22 @@ def _validate_types(self) -> None:
...
     def __post_init__(self) -> None:
         self._validate_types()
+
+
+ConfigType = TypeVar("ConfigType", bound=ModelConfiguration)
+ModuleType = TypeVar("ModuleType", bound=nn.Module)
+
+
+@dataclass
+class SubassemblyFactory(Generic[ConfigType, ModuleType]):
```
Comment (albertz) [inline i6_models/config.py:63]: Should this be versionized?
```suggestion
class SubassemblyFactoryV1(Generic[ConfigType, ModuleType]):
```

Comment (albertz) [inline i6_models/config.py:63]: I don't like the name so much. This is not really just for subassemblies. This can (and probably will) be used for any of our modules which have this generic interface where `__init__` accepts the config object.
```suggestion
class ModuleFactoryV1(Generic[ConfigType, ModuleType]):
```
?

File: i6_models/config.py (hunk 4 above)
Comment (albertz) [reply i6_models/config.py:69]: Btw, to make type hinting happy about this, you might need to adapt the `ModuleType` definition, do define it in a way that says that its `__init__` is in this way.

Comment (Atticus1806) [reply i6_models/config.py:69]: @SimBe195

File: i6_models/config.py (hunk 5 above)
Comment (Atticus1806) [reply i6_models/config.py:63]: I dont think so. This is no part/assembly/primitive

Comment (albertz) [reply i6_models/config.py]: So only parts/assemblies/primitives should be versionized, and everything else not? Why this inconsistency?

Comment (Atticus1806) [reply i6_models/config.py]: I don't have a strong opinion on this, but I think we don't expect this module to change, since its quite basic.

File: i6_models/config.py (hunk 4 above)
Comment (SimBe195) [reply i6_models/config.py:69]: > Btw, to make type hinting happy about this, you might need to adapt the ModuleType definition, do define it in a way that says that its __init__ is in this way.

And how would this work? Maybe by introducing a new protocol class? @albertz

Comment (albertz) [reply i6_models/config.py:69]: I haven't really tried it but I assume so.

Where exactly does it fail anyway? We could also just ignore the type checking for that line where it fails.

I don't think we should define a wrong type here just to make type checking happy. We should define the correct type, and fix the type checking, or disable type checking where it does not work.

Before disabling type checking for that line, I would first try to define the type in the correct way though.

Comment (Atticus1806) [reply i6_models/config.py:69]: So I just tested in my PyCharm and `Callable[[ModelConfiguration], ModuleType]` causes 
```
class Test(nn.Module):
    def __init__(self, cfg: TestConfig):
        super().__init__()
        self.x = nn.Linear(cfg.param, cfg.param)
```
to not be recognized as matching type while `Type[ModuleType]`.
When constructing in both cases I did not get a typeguard error.

Comment (SimBe195) [reply i6_models/config.py:69]: I currently do not see a way to give typehints for how we expect the `__init__` of a class to look like. But maybe what we could do is at least to put an assertion into `__post_init__` that checks that the signature of `self.module_class.__init__` is correct.

Comment (SimBe195) [reply i6_models/config.py:69]: > and Callable[[ModelConfiguration], ModuleType] causes [...] to not be recognized as matching type

It should be `Callable[[ConfigType], ModuleType]`. But anyway I believe `Type[ModuleType]` + a runtime assertion is preferable.

Comment (albertz) [reply i6_models/config.py:69]: > So I just tested in my PyCharm and `Callable[[ModelConfiguration], ModuleType]` ...

But my suggestion was to actually not use this and use `Type[ModuleType]` instead? I'm a bit confused now.

> to not be recognized as matching type while `Type[ModuleType]`.

I think there is a typo? I'm not sure I understand the sentence correctly.

---

> It should be `Callable[[ConfigType], ModuleType]`.

No I think it really should be `Type[ModuleType]`. That is what we expect here.

> But anyway I believe `Type[ModuleType]` + a runtime assertion is preferable.

What runtime assertion do you mean? The code I posted was just for demonstration, which would require that it is `Type[ModuleType]`, not `Callable[[ConfigType], ModuleType]`. It was just an example to demonstrate why `Type[ModuleType]` is the correct thing here. 

---



> > Btw, to make type hinting happy about this, you might need to adapt the ModuleType definition, do define it in a way that says that its **init** is in this way.
> 
> And how would this work? Maybe by introducing a new protocol class? @albertz

Actually I tried this code now and I don't see any warning? As I asked before, where exactly do you get a warning when you use `Type[ModuleType]`? I don't see this. I tried with [this code](https://github.com/albertz/playground/blob/master/module_init_typing.py) and I don't see any warnings.

Comment (SimBe195) [reply i6_models/config.py:69]: > But my suggestion was to actually not use this and use Type[ModuleType] instead? I'm a bit confused now.

There are basically two approaches:

1. `module_class: Callable[[ConfigType], ModuleType]` which I suggested initially to type-hint the constructor correctly but this has the disadvantage that you can't use things like `isinstance(obj, opt.module_class)` like you said.
2. `module_class: Type[ModuleType]` which is more how we want it but gives no typehints for the constructor, so calling `self.module_class(self.cfg)` may raise a type warning. @Atticus1806  just wanted to test both options in PyCharm.

> No I think it really should be Type[ModuleType]. That is what we expect here.

Yes I agree but only if it's still checked that the constructor of the given `module_class` has the right signature. This way it will already raise an error in sisyphus when you create a `ModelConfiguration` and not only later in the training job.

> What runtime assertion do you mean?

An assertion in `__post_init__` that checks the signature of `module_class.__init__` as I wrote in my [previous message](https://github.com/rwth-i6/i6_models/pull/21#discussion_r1242084866).

> Actually I tried this code now and I don't see any warning? As I asked before, where exactly do you get a warning when you use Type[ModuleType]? I don't see this. I tried with [this code](https://github.com/albertz/playground/blob/master/module_init_typing.py) and I don't see any warnings.

In VSCode (Pylance) do get a warning:

**Comment by Atticus1806** [comment]
@SimBe195  @albertz  is this fine for you? Added a test and the assertion check that the type matches.

File: i6_models/config.py [hunk 6]
```diff
@@ -51,3 +54,25 @@ def _validate_types(self) -> None:
...
+
+    module_class: Type[ModuleType]
+    cfg: ConfigType
+
+    def __call__(self) -> ModuleType:
+        """Constructs an instance of the given module class"""
+        return self.module_class(self.cfg)
+
+    def __post_init__(self) -> None:
+        assert signature(self.module_class).parameters["cfg"].annotation == type(self.cfg)
```
Comment (SimBe195) [inline i6_models/config.py:78]: We shouldn't force the parameter to be called "cfg". Rather check that there is a single parameter and check the type of that one.
```suggestion
        parameters = signature(self.module_class).parameters.values()
        assert len(parameters) == 1
        cfg_parameter = next(iter(parameters))
        assert typeguard.check_type(self.cfg, cfg_parameter.annotation)
```

**Review by SimBe195 — APPROVED** [review]

File: i6_models/config.py [hunk 7]
```diff
@@ -51,3 +54,28 @@ def _validate_types(self) -> None:
...
+
+    def __call__(self) -> ModuleType:
+        """Constructs an instance of the given module class"""
+        return self.module_class(self.cfg)
+
+    def __post_init__(self) -> None:
+        parameters = signature(self.module_class).parameters.values()
+        assert len(parameters) == 1
+        cfg_parameter = next(iter(parameters))
+        typeguard.check_type(self.cfg, cfg_parameter.annotation)
```
[the remark this answers is not in the document]
Comment (Atticus1806) [reply i6_models/config.py:81]: Can you give an example where this "fails"?

[the remark this answers is not in the document]
Comment (albertz) [reply i6_models/config.py:81]: I think in general using `signature` is not totally reliable and it is mostly intended to be used for inspection and debugging.

My examples were just examples. I can imagine there are more cases.

I just don't think such code is really necessary. I would keep it simple. But if you really want to have it, then keep it.

However, I think better than such a check would be to fix the typing, as I said before. Then PyCharm and other tools could also print you warnings.

[the remark this answers is not in the document]
Comment (albertz) [reply i6_models/config.py:81]: I don't know. I haven't really tried.

I just asked the [question on StackOverflow](https://stackoverflow.com/questions/76572323/combine-protocol-with-base-class-requirement-type-intersection). If we don't get a good answer there, the [official `typing` package on GitHub](https://github.com/python/typing) might be another good place to ask, or the [official `mypy` package on GitHub](https://github.com/python/mypy).

File: i6_models/config.py [hunk 8]
```diff
@@ -17,6 +17,9 @@ class ExampleModule(Module):
 from __future__ import annotations
 from dataclasses import dataclass, fields
 import typeguard
+from torch import nn
+from typing import Generic, TypeVar, Type
+from inspect import signature
```
Comment (SimBe195) [inline i6_models/config.py:22]:
```suggestion
import inspect
```

File: i6_models/config.py (hunk 7 above)
Comment (SimBe195) [inline i6_models/config.py:81]: Result of offline discussion: Solving the issue via type-hinting seems to be either not possible or overly complicated. Instead have a more sophisticated type-check here that should cover all cases of how `module_class` can be defined.
```suggestion
        # Check typing of module_class and cfg, i.e. make sure that "self.module_class(self.cfg)" is a valid call.
        parameters = inspect.signature(self.module_class).parameters.values()
        assert len(parameters) >= 1
        parameter_iter = iter(parameters)

        # 1. Check that the first parameter is either not annotated or the annotation matches the type of self.cfg
        cfg_parameter = next(parameter_iter)
        if cfg_parameter.annotation is not inspect.Parameter.empty:
            typeguard.check_type(self.cfg, cfg_parameter.annotation)
            
        # 2. Check that all other parameters have default values
        for parameter in parameter_iter:
            assert parameter.default is not inspect.Parameter.empty
```

**Review by SimBe195 — CHANGES_REQUESTED** [review]

**Review by SimBe195 — APPROVED** [review]

File: i6_models/config.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (SimBe195) [reply i6_models/config.py]: Yes we decided that we want this check so that any error is detected early on. The check has been made a bit more general now so that all (or at least most) use-cases should be covered. If this breaks for some specific module class later we can adjust the check.

**Review by albertz — APPROVED** [review]
